### PR TITLE
Swerve drive modes

### PR DESCRIPTION
**Pull request by quinnmp**
This is my basic implementation and is **definitely incorrect**, specifically in `kinematic_common_interface.cpp`. Also, there are not yet requests to change wheel angles.

### CODE REVIEW

File: src/world_interface/kinematic_common_interface.cpp
Comment (quinnmp) [inline src/world_interface/kinematic_common_interface.cpp:149]: This method does not handle rotation correctly at the moment. I wasn't sure how to, because we want this to essentially act as the normal drive mode but with a rotated reference frame

File: src/Constants.h
Comment (abhaybd) [inline src/Constants.h:105]: These should be calculated from the swerve kinematics, not hardcoded, since the angles will change based on the wheelbase dimensions.

File: src/network/MissionControlProtocol.cpp
Comment (abhaybd) [inline src/network/MissionControlProtocol.cpp:101]: I think you mean `for (int i = 0; i < 4; i++)`

File: src/network/MissionControlProtocol.cpp
Comment (abhaybd) [inline src/network/MissionControlProtocol.cpp:117]: I know we do this in a lot of places, but why do we `dump()` and then don't use `msg`?

File: src/network/MissionControlProtocol.cpp
Comment (abhaybd) [inline src/network/MissionControlProtocol.cpp:186]: Super minor, but `std::hypot` is more concise

File: src/world_interface/kinematic_common_interface.cpp
Comment (abhaybd) [inline src/world_interface/kinematic_common_interface.cpp:63]: Same as above, 0-indexing

File: src/world_interface/kinematic_common_interface.cpp
Comment (abhaybd) [inline src/world_interface/kinematic_common_interface.cpp:121]: Index out of bounds

**Review by abhaybd — COMMENTED** [review]
Brief review, didn't get a chance to look at everything yet

File: src/network/MissionControlProtocol.cpp
Comment (quinnmp) [reply src/network/MissionControlProtocol.cpp:117]: I actually don't know - it was done in `validateDriveRequest` and so I emulated it in the others. I'll remove them for now and revert if needed

File: src/Constants.h
Comment (abhaybd) [inline src/Constants.h:107]: Can we put this in a nested namespace, so maybe it would be `Constants::Drive` or something?

File: src/Globals.h
Comment (abhaybd) [inline src/Globals.h:37]: Can we add documentation on what the parts of the drive mode are? Particularly the `bool`, which is vague

Comment (abhaybd) [inline src/Globals.h:37]: Also, since there's just two, we should probably just use `std::pair`

File: src/Constants.h
Comment (abhaybd) [inline src/Constants.h:103]: Just a note, these motors are the actual wheels, we should add `motorid_t` values for the steer motors. I think Victor should handle that in his PR, though, so wait for that and you can merge it in

File: src/Constants.cpp
Comment (abhaybd) [inline src/Constants.cpp:8]: We should probably reuse `robot::swerveKinematics()` instead of re-constructing the swerve kinematics

**Review by abhaybd — CHANGES_REQUESTED** [review]
Pretty good! Just some notes, and we should be able to start testing this after Victor's pr

File: src/Globals.cpp
Comment (huttongrabiel) [inline src/Globals.cpp:43]: Same thing here as Abhay said. `std::pair` instead of `std::tuple` if you make that change in `Globals.h`.

File: src/world_interface/kinematic_common_interface.cpp
Comment (huttongrabiel) [inline src/world_interface/kinematic_common_interface.cpp:133]: Is there an easy way to factor this out into a little function? It's used in 3 places throughout. No big deal if not

File: src/world_interface/world_interface.h
Comment (huttongrabiel) [inline src/world_interface/world_interface.h:102]: Did you mean frame instead of from?

**Review by huttongrabiel — CHANGES_REQUESTED** [review]

File: src/Constants.h
Comment (abhaybd) [inline src/Constants.h:16]: Looks unused. Also btw, `using` declarations in header files should usually be avoided, since it causes namespace pollution.

File: src/Globals.cpp
Comment (abhaybd) [inline src/Globals.cpp:13]: Unused?

File: src/Globals.h
Comment (abhaybd) [inline src/Globals.h:26]: Since we include the header file, we don't need the forward declaration. Also the other stuff is unused?

File: src/control/SwerveController.cpp
Comment (abhaybd) [inline src/control/SwerveController.cpp:24]: Do you mean `dtheta == 0`?

File: src/control/SwerveController.h
Comment (abhaybd) [inline src/control/SwerveController.h:78]: Can we make this doxy-style comments

File: src/control/SwerveController.h
Comment (abhaybd) [inline src/control/SwerveController.h:18]: Instead of a global map, I think it would be a bit cleaner to provide an overload of the [`to_string()`](https://github.com/huskyroboticsteam/Resurgence/blob/master/src/utils/core.h#L50-L53) function.

To do that, do something like
```c++
namespace util {
std::string to_string(control::DriveMode mode);
}
```

See an example [here](https://github.com/huskyroboticsteam/Resurgence/blob/master/src/world_interface/data.h#L305-L307)

PS: `static` globals and methods in header-files are kind of a code smell. It will just be compiled separately everywhere the header is included, and all it does is prevent a [ODR](https://en.cppreference.com/w/cpp/language/definition) violation. If we wanted to put this map here, we should declare it as `extern const` and provide the definition in `SwerveController.cpp`

File: src/control/SwerveController.h
Comment (abhaybd) [inline src/control/SwerveController.h:33]: Let's use `power` instead of `pwm`, since the exact method of setting the power is an implementation detail. For example, BLDC motors don't even use PWM.

File: src/control/SwerveController.h
Comment (abhaybd) [reply src/control/SwerveController.h:65]: Also, this (and most methods in this class) should be marked as `const`

File: src/control/SwerveController.h
Comment (abhaybd) [inline src/control/SwerveController.h:79]: Now that this is encapsulated in a class, it lessens its readability to package this in a pair. We can just separate these out into class fields. Also, public fields are ugly and kind of leak implementation, so maybe make this private and add appropriate getters and setters.

Comment (abhaybd) [reply src/control/SwerveController.h:79]: Also, maybe the `setDriveMode` method can also return the corresponding `swerve_rots_t`, which makes client code a little more elegant

File: src/control/SwerveController.h
Comment (abhaybd) [reply src/control/SwerveController.h:81]: Also to be more in line with the rest of the class, the type should probably be `swerve_rots_t`

File: src/network/MissionControlTasks.cpp
Comment (abhaybd) [inline src/network/MissionControlTasks.cpp:116]: This is a lot of duplicated code, maybe factor the common parts out?

File: src/world_interface/world_interface.h
Comment (abhaybd) [inline src/world_interface/world_interface.h:23]: Unused?

File: src/world_interface/world_interface.h
Comment (abhaybd) [inline src/world_interface/world_interface.h:91]: Parameter names should be `left`, `right`

File: src/Globals.cpp
Comment (abhaybd) [inline src/Globals.cpp:53]: /We should use the real wheel base instead of the effective wheel base. If `ROBOT_WIDTH` isn't already a constant, we should add it

File: src/CMakeLists.txt
Comment (abhaybd) [inline src/CMakeLists.txt:155]: This should probably be added to the `control` target instead of `utils`

**Review by abhaybd — CHANGES_REQUESTED** [review]
Great job on this! I admittedly got a little carried away with the review comments, but this is a nice first draft, and the core functionality and design is all great, it's just cleanliness and QoL fixes at this point.

File: src/Constants.h
Comment (abhaybd) [inline src/Constants.h:93]: This should be defined as `extern`, in keeping with everything else. Also, can you document what this constant is? (what it is, what the units are, etc)

File: src/control/SwerveController.cpp
Comment (abhaybd) [inline src/control/SwerveController.cpp:66]: No point in checking the `E_STOP` value, since this method isn't responsible for policing that. Also means we can cut the dependency on globals

Comment (abhaybd) [reply src/control/SwerveController.cpp:66]: As an aside, we can just cut this whole special case, since if the inputs are zero then the outputs will be zero.

File: src/network/MissionControlProtocol.cpp
Comment (abhaybd) [inline src/network/MissionControlProtocol.cpp:108]: Can we log a warning and no-op if `target` is uninitialized? i.e. if the mode was the wrong value. In theory this shouldn't happen bc of the validator, but a buggy validator shouldn't be able to damage the hardware if some uninitialized value is passed as the pid target.

File: src/network/MissionControlProtocol.cpp
Comment (abhaybd) [reply src/network/MissionControlProtocol.cpp:46]: Makes sense, what if we just changed the methods of the power repeat task to also execute it now? That way we can just call `setTurnInPlaceCmdVel` or whatever and it will both schedule the power repeat *and* send the command now. That way the utility functions can only exist in MCPTasks

File: src/network/MissionControlTasks.cpp
Comment (abhaybd) [inline src/network/MissionControlTasks.cpp:33]: Pass structs by const ref!

File: src/world_interface/kinematic_common_interface.cpp
Comment (abhaybd) [inline src/world_interface/kinematic_common_interface.cpp:45]: We should handle the case where any of these datapoints are invalid, otherwise it will crash if we haven't gotten an encoder packet yet. I added a PR (#324) that should make things easier so we can just add a try-catch around the whole block.

Comment (abhaybd) [reply src/world_interface/kinematic_common_interface.cpp:45]: Note that this should be done everywhere else we do a `getData()` without checking if the datapoint is valid.

**Review by abhaybd — CHANGES_REQUESTED** [review]
Looking better, just some notes and some minor stuff I missed before.

File: src/network/MissionControlTasks.cpp
Comment (quinnmp) [inline src/network/MissionControlTasks.cpp:252]: Should this be  `bad_datapoint_access` instead of `std::runtime_error` once that PR is merged?

Comment (abhaybd) [reply src/network/MissionControlTasks.cpp:252]: Yessir

File: src/world_interface/kinematic_common_interface.cpp
Comment (abhaybd) [inline src/world_interface/kinematic_common_interface.cpp:47]: Same as above, remember to switch this to `bad_datapoint_access`

File: src/world_interface/kinematic_common_interface.cpp
Comment (abhaybd) [inline src/world_interface/kinematic_common_interface.cpp:82]: Same as above you get the gist

**Review by abhaybd — COMMENTED** [review]
Looks good! Let's get this tested asap (after setting up hardware) and then this should be good to merge

File: src/world_interface/world_interface.h
Comment (abhaybd) [inline src/world_interface/world_interface.h:92]: This declaration should go in `control_interface.h` I think

**Review by abhaybd — CHANGES_REQUESTED** [review]
This looks great! One minor nitpick but besides that we should be ready to test soon!

File: src/network/MissionControlProtocol.cpp
Comment (imisaacwu) [inline src/network/MissionControlProtocol.cpp:104]: These should be setMotorPos I think

Comment (quinnmp) [reply src/network/MissionControlProtocol.cpp:104]: Good catch!

**Review by abhaybd — APPROVED** [review]
Tested and approved! Great job on this!

**Pull request merged**
